Sources/NIOCore/ByteBuffer-lengthPrefix.swift

```diff
@@ -0,0 +1,105 @@
+        writeMessage: (inout ByteBuffer) throws -> Int
+    ) throws -> Int where Integer: FixedWidthInteger {
+        var totalBytesWritten = 0
+        
+        let lengthPrefixIndex = self.writerIndex
+        // Write a zero as a placeholder which will later be overwritten by the actual number of bytes written
+        totalBytesWritten += self.writeInteger(.zero, endianness: endianness, as: Integer.self)
+        
+        let startWriterIndex = writerIndex
+        let messageLength = try writeMessage(&self)
+        let endWriterIndex = writerIndex
+        
+        totalBytesWritten += messageLength
+        
+        let actualBytesWritten = endWriterIndex - startWriterIndex
+        assert(
+            actualBytesWritten == messageLength, 
+            "writeMessage returned \(messageLength) bytes, but actually \(actualBytesWritten) bytes were written, but they should be the same"
+        )
+        
+        guard let lengthPrefix = Integer(exactly: messageLength) else {
+            throw LengthPrefixError.messageLengthDoesNotFitExactlyIntoRequiredIntegerFormat
+        }
+        
+        self.setInteger(lengthPrefix, at: lengthPrefixIndex, endianness: endianness, as: Integer.self)
+        
+        return totalBytesWritten
+    }
+}
+
+extension ByteBuffer {
+    /// Reads an `integer` from `self` and passes it to `readMessage`. 
+    /// It is checked that exactly the number of bytes specified in the length prefix is read during the call to `readMessage`. 
+    /// 
+    /// If nil is returned, `readerIndex` is **not** moved forward.
+    /// - Parameters:
+    ///     - endianness: The endianness of the length prefix `Integer` in this `ByteBuffer` (defaults to big endian).
+    ///     - integer: the desired `Integer` type used to read the length prefix
+    ///     - readMessage: A closure that takes the message length, reads exactly that number of bytes from the given `ByteBuffer` and returns the result.
+    /// - Returns: `nil` if the length prefix could not be read, 
+    ///            the length prefix is negative or
+    ///            the buffer does not contain enough bytes to read a message of this length.
+    ///            Otherwise the result of `readMessage`.
+    @inlinable
+    public mutating func readLengthPrefixed<Integer, Result>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        readMessage: (Int, inout ByteBuffer) throws -> Result?
+    ) rethrows -> Result? where Integer: FixedWidthInteger {
+        let prefixSize = MemoryLayout<Integer>.size
+        
+        guard let lengthPrefix = self.getInteger(at: readerIndex, endianness: endianness, as: Integer.self),
+              let messageLength = Int(exactly: lengthPrefix),
+              var messageBuffer = self.getSlice(at: readerIndex + prefixSize, length: messageLength)
+        else {
+            return nil
+        }
+        
+        guard let result = try readMessage(messageLength, &messageBuffer) else {
```

Comment:
Do we want to pass in the message length? It can be recovered from the slice if necessary.

Reply:
Also wondering if we should drop the `inout` when passing in the slice, I think it makes it clearer that "this is your slice, do what you want with it".

Reply:
The "Problem" is that you would need to make your own local copy to actually read something from that slice.
This would no longer work:
```swift
buffer.readLengthPrefixed(as: UInt8.self) { buffer in 
    buffer.readString(buffer.readableBytes)
}
```
You would need to write the following:
```swift
buffer.readLengthPrefixed(as: UInt8.self) { buffer in 
    var buffer = buffer
    return buffer.readString(buffer.readableBytes)
}
```

I'm also thinking about removing the `readLengthPrefixed` method and just keeping `readLengthPrefixedSlice` because you could just write something like this:
```swift
guard var slice = buffer.readLengthPrefixedSlice(as: UInt8.self),
           let message = slice.readString(slice.readableBytes) 
else {
    throw MyCustomError()
}
```
At least in the case of RSocket we probably want to throw a custom error anyway. WDYT?

Reply:
I think I'd be fine with that.

Reply:
Same here.